mitiq/typing.py

```diff
@@ -115,11 +124,47 @@ def nqubits(self) -> int:
     def asarray(self) -> npt.NDArray[np.int64]:
         return self._bitstrings
 
-    def __getitem__(self, indices: List[int]) -> npt.NDArray[np.int64]:
-        return np.array([self._measurements[i] for i in indices]).T
-
-    def __iter__(self) -> Iterable[Bitstring]:
-        yield from self.result
+    @classmethod
+    def from_counts(
+        cls,
+        counts: Dict[str, int],
+        qubit_indices: Optional[Tuple[int, ...]] = None,
+    ) -> "MeasurementResult":
+        """Initializes a MeasurementResult from a dictionary of counts.
+
+        Example:
+        >>> MeasurementResult.from_counts({"00": 175, "11": 177})
+        """
+        counter = Counter(counts)
+        return cls(list(counter.elements()), qubit_indices)
+
+    def get_counts(self) -> Dict[str, int]:
+        """Returns a colleciton.Counter whose keys are the measured
+        bitstrings the and whose values are the counts.
+        """
+        strings = ["".join(map(str, bits)) for bits in self.result]
+        return dict(Counter(strings))
+
+    def to_dict(self) -> Dict[str, Any]:
+        """Returns a Python dictionary whose keys are the measured
+        bitstrings the and whose values are the counts.
+
+        Note: Qubit indices (self.qubit_indeces) are lost in the conversion.
+        """
+        data = {
+            "qubit_indices": self.qubit_indices,
+            "shots": self.shots,
+            "counts": self.get_counts(),
+        }
+        return data
+
+    def filter_qubits(self, qubit_indices: List[int]) -> npt.NDArray[np.int64]:
+        """Returns the bitstrings associated to a subset of qubits."""
+        return np.array([self._measurements[i] for i in qubit_indices]).T
```

Comment:
Is there a reason you like making a custom method, over the previously used `getitem` way? I don't have a preference, just curious if there is a reason.

Reply:
The reason is that `getitem` is very natural for arrays with contiguous and unambiguous indices. 
In our case the function is used and defined for `qubit_indeces` and so its behavior would be quite ambiguous. 

E.g. is `result[3] ` the third measurement result, the results of the third bit, or the results of the qubit with index 3?

So, I preferred to use a different and more explicit name.

Reply:
Gotcha, that makes sense. Thanks for the explanation, and I agree this seems like the right way to go.